### PR TITLE
Add new fields in Credit form

### LINKED ISSUE #329: Add new "Credit card" account form
We're working on adding each new account type as they've got slightly different "field" needs.

All accounts have "Name" and "Balance" by default.

Additional fields:
* Type — Select (Mastercard, Visa, American Express, Discover)
* Limit — Number

View is in `/views/accounts/account/_credit.html.erb`

### PR DESCRIPTION
**Pull request by sarvaiyanidhi**
Closes #329 

This PR address below changes
- New field 'limit' in accounts table
-  Credit type of account captures type and limit details

### CODE REVIEW

**Review by OohBen — APPROVED** [review]

File: db/migrate/20240207021033_add_limit_to_accounts.rb [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class AddLimitToAccounts < ActiveRecord::Migration[7.2]
+  def change
+    add_column :accounts, :limit, :bigint, default: 0
```
Comment (rmarescu) [inline db/migrate/20240207021033_add_limit_to_accounts.rb:3]: Is this field meant to store money? If so, should use the `limit_cents` convention, as per a [previous discussion](https://github.com/maybe-finance/maybe/pull/268)?

File: app/views/accounts/account/_credit.html.erb [hunk 2]
```diff
@@ -0,0 +1,8 @@
+<div class="relative p-4 border border-gray-100 bg-offwhite rounded-xl focus-within:bg-white focus-within:shadow focus-within:opacity-100">
+  <%= f.label :subtype, t('.type') ,class: 'block text-sm font-medium opacity-50 focus-within:opacity-100'  %>
```
Comment (rmarescu) [inline app/views/accounts/account/_credit.html.erb:2]: Nit - to match the surrounding code

```suggestion
  <%= f.label :subtype, t(".type") ,class: "block text-sm font-medium opacity-50 focus-within:opacity-100"  %>
```

File: app/views/accounts/account/_credit.html.erb [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+  <%= f.select :subtype, options_for_select(Account::Credit::SUBTYPES, selected: ""), {}, class: "block w-full p-0 mt-1 bg-transparent border-none focus:outline-none focus:ring-0" %>
+</div>
+<div class="relative p-4 border border-gray-100 bg-offwhite rounded-xl focus-within:bg-white focus-within:shadow focus-within:opacity-100">
+  <%= f.label :limit, t('.limit') ,class: 'block text-sm font-medium opacity-50 focus-within:opacity-100'  %>
```
Comment (rmarescu) [inline app/views/accounts/account/_credit.html.erb:6]:
```suggestion
  <%= f.label :limit, t(".limit") ,class: "block text-sm font-medium opacity-50 focus-within:opacity-100" %>
```

File: app/views/accounts/account/_credit.html.erb [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+  <%= f.number_field :limit, placeholder: number_to_currency(0), in: 0.00..100000000.00, required: 'required', class: "p-0 mt-1 bg-transparent border-none opacity-50 focus:outline-none focus:ring-0 focus-within:opacity-100" %>
```
Comment (rmarescu) [inline app/views/accounts/account/_credit.html.erb:7]:
```suggestion
  <%= f.number_field :limit, placeholder: number_to_currency(0), in: 0.00..100000000.00, required: "required", class: "p-0 mt-1 bg-transparent border-none opacity-50 focus:outline-none focus:ring-0 focus-within:opacity-100" %>
```

File: app/models/account/credit.rb [hunk 5]
```diff
@@ -1,3 +1,9 @@
 class Account::Credit < ApplicationRecord
   include Accountable
+  SUBTYPES = [
```
Comment (rmarescu) [inline app/models/account/credit.rb:3]: Could `SUBTYPES` be an array, and the human-friendly version defined as translations?

```ruby
  SUBTYPES = %w[ mastercard visa american_express discover ]
```

File: app/views/accounts/account/_credit.html.erb (hunk 4 above)
Comment (rmarescu) [inline app/views/accounts/account/_credit.html.erb:7]: If `limit` must be positive, should add some model validation to enforce it.

File: app/controllers/accounts_controller.rb [hunk 6]
```diff
@@ -30,7 +30,7 @@ def create
   private
 
   def account_params
-    params.require(:account).permit(:name, :accountable_type, :balance, :balance_cents, :subtype)
+    params.require(:account).permit(:name, :accountable_type, :balance, :balance_cents, :subtype, :limit)
```
Comment (rmarescu) [inline app/controllers/accounts_controller.rb:33]: Could the change have some test coverage?

**Review by rmarescu — COMMENTED** [review]
@sarvaiyanidhi I've added some comments for consideration!

File: app/views/accounts/account/_credit.html.erb (hunk 2 above)
Comment (pmareke) [reply app/views/accounts/account/_credit.html.erb:2]: Good catch! I think we should think about in adding Rubocop asap and detect this kind of errors automatically.

File: db/migrate/20240207021033_add_limit_to_accounts.rb (hunk 1 above)
Comment (sarvaiyanidhi) [reply db/migrate/20240207021033_add_limit_to_accounts.rb:3]: I was not sure that limit was also supposed to be money field. @Shpigford Can you confirm if we want to monetize limit field as well which will generate limit_cents and limit_currency in accounts table.

Comment (rmarescu) [reply db/migrate/20240207021033_add_limit_to_accounts.rb:3]: I've noticed that #268 introduced `balance_currency` (associated with `balance_cents`. Is there a case where there is a need for separate currency fields for each money fields? Like the balance be in USD, but the limit in CAD for example?

Is there a way to have just one currency field, that can be shared by all money fields?

Comment (sarvaiyanidhi) [reply db/migrate/20240207021033_add_limit_to_accounts.rb:3]: Yeah, even I was thinking on same. Even there is separate 'currency' field already present in accounts table. Not sure if we still need it as balance_currency is also added in table.

**Comment by Shpigford** [comment]
Closing for new per the comment here: https://github.com/maybe-finance/maybe/issues/329#issuecomment-1932637291

**Pull request closed without merging**